File: frontend/src/services/projectDistribution.ts
import api from './api'

export interface DistAttachment {
  id: number
  distribution_id: number
  category?: string          // 附件 / 审签表
  original_name: string
  saved_name: string
  file_size: number
  mime_type: string
  uploaded_by: string
  uploaded_at: string
}

export interface Distribution {
  id: number
  serial_no: string
  originator: string
  form_type: string
  name: string
  content: string
  budget: number | null
  price_limit: number | null
  method: string
  org_form: string
  manage_dept: string
  demand_dept: string
  project_number: string
  extra: string          // 各流程专有字段(JSON字符串)
  is_central: number
  officer: string
  agency_code: string
  agency_name: string
  source: string
  status: string
  project_id: number | null
  created_by: string
  created_at: string
  updated_at: string
  attachments: DistAttachment[]
}

export const listDistributions = () =>
  api.get<{ ok: boolean; data: Distribution[] }>('/distributions')

export const createDistribution = (data: Partial<Distribution>) =>
  api.post<{ ok: boolean; data: Distribution }>('/distributions', data)

export const updateDistribution = (id: number, data: Partial<Distribution>) =>
  api.put<{ ok: boolean; data: Distribution }>(`/distributions/${id}`, data)

export const deleteDistribution = (id: number) =>
  api.delete<{ ok: boolean }>(`/distributions/${id}`)

export const reassignAgency = (id: number, agency_code?: string) =>
  api.post<{ ok: boolean; data: Distribution }>(`/distributions/${id}/reassign-agency`, { agency_code })

export const uploadDistAttachment = (did: number, file: File) => {
  const fd = new FormData()
  fd.append('file', file)
  return api.post(`/distributions/${did}/attachments`, fd, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
}

export const deleteDistAttachment = (did: number, aid: number) =>
  api.delete(`/distributions/${did}/attachments/${aid}`)

export const distAttachmentPreviewUrl = (did: number, aid: number) =>
  `/api/distributions/${did}/attachments/${aid}/preview`

export const distAttachmentDownloadUrl = (did: number, aid: number) =>
  `/api/distributions/${did}/attachments/${aid}/download`

// 合并选中的 PDF 为一个 PDF（一键打印）
export const distPrintUrl = (did: number, ids: number[]) =>
  `/api/distributions/${did}/print?ids=${ids.join(',')}`

export const scrapeRdweb = () =>
  api.post<{ ok: boolean; message?: string; error?: string }>('/distributions/scrape-rdweb')

export const scrapeStatus = () =>
  api.get<{ ok: boolean; data: { running: boolean; last_msg: string } }>('/distributions/scrape-status')

// ── rd-web 办理动作（接收/驳回/撤回，后端 Playwright RPA，异步）──
export const rdwebAction = (
  did: number, action: 'accept' | 'reject' | 'withdraw', officer?: string, opinion?: string,
) =>
  api.post<{ ok: boolean; message?: string; error?: string }>(
    `/distributions/${did}/rdweb-action`, { action, officer, opinion })

export const rdwebActionStatus = () =>
  api.get<{ ok: boolean; data: { running: boolean; last_msg: string; ok: boolean | null; action?: string; serial?: string } }>(
    '/distributions/rdweb-action-status')

// ── rd-web 账号维护 ──
export interface RdwebAccount {
  id: number
  owner: string
  phone: string
  password: string
  usage: string     // 分发 / 执行
  note: string
  updated_at: string
}
export const listRdwebAccounts = () =>
  api.get<{ ok: boolean; data: RdwebAccount[] }>('/distributions/rdweb-accounts')
export const createRdwebAccount = (data: Partial<RdwebAccount>) =>
  api.post('/distributions/rdweb-accounts', data)
export const updateRdwebAccount = (id: number, data: Partial<RdwebAccount>) =>
  api.put(`/distributions/rdweb-accounts/${id}`, data)
export const deleteRdwebAccount = (id: number) =>
  api.delete(`/distributions/rdweb-accounts/${id}`)

// 导出 Excel（按时间范围/采购方式[多选]筛选）。返回可直接下载的 URL。
export const distExportUrl = (p: { date_from?: string; date_to?: string; methods?: string[] }) => {
  const qs = new URLSearchParams()
  if (p.date_from) qs.set('date_from', p.date_from)
  if (p.date_to) qs.set('date_to', p.date_to)
  if (p.methods && p.methods.length) qs.set('methods', p.methods.join(','))
  return `/api/distributions/export?${qs.toString()}`
}

File: backend/models/project_distribution.py
from . import db


class ProjectDistribution(db.Model):
    """采购部助理分发的项目 = 项目立项的起点。

    信息来源：rd-web（医互通）抓取 或 手动录入。仅「指定经办人 + 采购部助理」可见。
    经办人立项时自动带入 名称/内容/预算（可编辑）；附件流转到采购需求编制可直接用。
    """
    __tablename__ = "project_distributions"

    id = db.Column(db.Integer, primary_key=True)
    serial_no = db.Column(db.String(100), default="")     # rd-web 采购需求流水号
    originator = db.Column(db.String(50), default="")     # 发起人（rd-web 流程发起人）
    form_type = db.Column(db.String(40), default="采购需求审签表")  # 来源表单类型（多流程区分）
    name = db.Column(db.String(300), default="")          # 项目名称
    content = db.Column(db.Text, default="")              # 项目内容
    budget = db.Column(db.Float)                          # 预算金额（元）
    price_limit = db.Column(db.Float)                     # 限价金额（元）
    method = db.Column(db.String(50), default="")         # 采购方式（院内竞选/政府采购/...）
    org_form = db.Column(db.String(50), default="")       # 采购组织形式（自行采购/委托代理...）
    manage_dept = db.Column(db.String(100), default="")   # 归口管理科室
    demand_dept = db.Column(db.String(100), default="")   # 需求科室
    project_number = db.Column(db.String(100), default="")# 项目编号（rd-web）
    extra = db.Column(db.Text, default="")                # 各流程专有字段(JSON)：照搬 rd-web 该表单全部字段
    is_central = db.Column(db.Integer, default=0)         # 是否政采中心项目（集采/医疗设备）
    officer = db.Column(db.String(50), default="")        # 指定经办人（display_name）
    agency_code = db.Column(db.String(10), default="")    # 指定代理机构（按规则自动派）
    source = db.Column(db.String(20), default="手动")      # 来源：rd-web / 手动
    status = db.Column(db.String(20), default="待分发")    # 待分发 / 已分发 / 已立项
    project_id = db.Column(db.Integer)                    # 立项后回填关联项目 id
    created_by = db.Column(db.String(50), default="")
    created_at = db.Column(db.String(30), default="")
    updated_at = db.Column(db.String(30), default="")

    def to_dict(self):
        return {c.name: getattr(self, c.name) for c in self.__table__.columns}


class RdwebAccount(db.Model):
    """rd-web（医互通）登录账号维护。账号会变，给陈梦霞(分发抓取)和经办人(执行)自己维护。
    usage='分发' 的账号被抓取程序使用（保存时同步到 SysConfig rdweb_loginuser/password）。"""
    __tablename__ = "rdweb_accounts"

    id = db.Column(db.Integer, primary_key=True)
    owner = db.Column(db.String(50), default="")          # 姓名/用途
    phone = db.Column(db.String(30), default="")          # 登录手机号
    password = db.Column(db.String(100), default="")      # 登录密码
    usage = db.Column(db.String(20), default="执行")       # 分发 / 执行
    note = db.Column(db.String(200), default="")
    updated_at = db.Column(db.String(30), default="")

    def to_dict(self):
        return {c.name: getattr(self, c.name) for c in self.__table__.columns}


class ProjectDistributionAttachment(db.Model):
    __tablename__ = "project_distribution_attachments"

    id = db.Column(db.Integer, primary_key=True)
    distribution_id = db.Column(db.Integer, nullable=False, index=True)
    category = db.Column(db.String(20), default="附件")    # 附件 / 审签表（流程打印PDF）
    original_name = db.Column(db.String(200))
    saved_name = db.Column(db.String(200))                # UUID 防冲突
    file_size = db.Column(db.Integer, default=0)
    mime_type = db.Column(db.String(100), default="")
    uploaded_by = db.Column(db.String(50), default="")
    uploaded_at = db.Column(db.String(30), default="")

    def to_dict(self):
        return {c.name: getattr(self, c.name) for c in self.__table__.columns}
